ASoC: SOF: ipc4-control: Add support for bytes control type

ujfalusi opened the pull request:
Hi,

RFC state atm due to few uncertainty regarding to the `struct sof_ipc4_abi_hdr` and the definition of the blob_type number in there.
There is also the `header.numid` in the TLV (bytes_ext) which in ipc3 was basically was set to `SOF_CTRL_CMD_BINARY` for all binary controls, but it is not clear what it should be with IPC4 and how that is used. This brings up some questions regarding to validation of the blob, is it sent for the correct module, and is it generic identification so that it can be used on different systems (not runtime determined 'random' number). Perhaps we should have UUID in the ABI header for IPC4 blobs like these?

the series implements support for MOD_INIT_INSTANCE (blob_type = 0) and MOD_LARGE_CONFIG_SET (blob_type = 4) type of blobs and tested with the later using FIR module.

ujfalusi commented:
Changes since v1:
- rebased on topic/sof-dev

singalsu reviewed (dismissed):
Thanks Peter! This PR works perfectly with FIR EQ component's byte control set/get (patches needed for FIR, topology, rimage).

ujfalusi commented:
Changes since v2:
- fixup patches to remove the `struct sof_ipc4_abi_hdr` and re-use/extend the existing `struct sof_abi_hdr` for IPC4 also.
- fixup to not store the binary blob under the widget as well since it should not be sent at mod_init_instance time.
- the binary control support is updated accordingly to the ABI header change.

@singalsu, @juimonen if you still have the blob you have tested with EQ then the only change needed for you is to move the size back where it is with IPC3 (3rd u32 instead the 2nd u32). Afaik the EQ module in firmware does not use the param_id, so leave the 2nd u32 as 0 to be safe (param_id = 0).

ranj063 commented on include/uapi/sound/sof/header.h:
```diff
@@ -25,35 +31,21 @@
 	__u32 data[];
 }  __packed;
 
-/**
```
if we're removing this anyway, why bother with the first patch at all?

ujfalusi replied:
We are not removing it, but for upstream this will be the first patch, then the extending to IPC4.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -893,18 +890,6 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 	if (process->payload_with_output_fmt)
```
@ujfalusi while you are at it. Can you please also fix this part? Looks like init instance with output format isnt supported either right? The wov module needs this and will just have to adapt to use the base_cfg_ext right?

ujfalusi replied:
Right, so another fixup to drop the whole token and flag?

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1646,6 +1648,29 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 		cfg += sizeof(struct sof_ipc4_audio_format);
 	}
 
+	fw_module = swidget->module_info;
```
do we need to do this in prepare? Can be done in setup no?

ujfalusi replied:
In setup we do not have the control_data, I think, so we need to do this here (also other type do it in there prepare)

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -194,19 +392,25 @@ static int sof_ipc4_widget_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_s
+				break;
+			case SND_SOC_TPLG_CTL_BYTES:
+				ret = sof_ipc4_set_bytes_data(sdev, scontrol, false);
+				if (ret < 0)
```
can the check go out side switch so we dont duplicate in both cases

ujfalusi replied:
sure.

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -181,11 +181,209 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
+	struct snd_ctl_tlv __user *tlvd = (struct snd_ctl_tlv __user *)binary_data;
+	struct sof_ipc4_control_data *cdata = scontrol->ipc_control_data;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_abi_hdr *data = cdata->data;
```
what if a kcontrol has no data in topology? cdata->data will be NULL until the user does a kcontrol set isnt it? In that case, dont we actually need to do a GET_LARGE_CONFIG to fetch the data from the DSP? And for volatile kcontrols we should always fetch this data from the DSP right?

ujfalusi replied:
I see, hrm, I see that we are fiddling with things in IPC3 side. Let me see how it can be done, but if the cdata->data is NULL then we do not have binary support in the first place, we need data to be allocated already.
If you mean that the content is 0, yeah, that might be the case, I have not implemented the volatile in this initial PR, but it is planed for later time.

ujfalusi commented:
Changes since v3:
- Drop payload_with_output_fmt and SOF_TKN_PROCESS_PAYLOAD_WITH_OUTPUT_FMT in fixups
- implement bytes_ext_volatile_get
- Move the error print outside of the switch to de-duplicate it in sof_ipc4_widget_kcontrol_setup()

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -181,32 +181,260 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
+	/* get all the component data from DSP */
+	if (from_dsp) {
+		struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+		int ret = sof_ipc4_set_bytes_data(sdev, scontrol, true);
```
is this correct @ujfalusi . Why do we set when we're doing a get?

aiChaoSONG commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1661,29 +1623,27 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 			continue;
+
 		sbe = (struct soc_bytes_ext *)kc->private_value;
 		scontrol = sbe->dobj.private;
```
We are looping the asoc widgets' kcontrol list and doing dereference directly, Pierre questions if it is safe before:https://github.com/thesofproject/linux/pull/4121#discussion_r1068580516, isn't it still a problem now?

ujfalusi replied:
If the kcontrol type is SND_SOC_TPLG_TYPE_BYTES we know exactly what is behind the private_value, but I will check if this can be made cleaner.

ujfalusi replied:
This part is now gone.

aiChaoSONG reviewed (commented):
I got a chance to test this PR for smart amp. we are able to send configuration blob with this one.

ujfalusi commented:
Changes since v4:
- Drop the rest of the references to `SOF_PROCESS_TOKENS` in the process widget fixup to be able to probe when there is a process widget in topology
- New function to set the module_id in the message template for the kcontrols associated with a widget
- Use the new helper to set the module_id for the message template associated with a process  module (we could have multiple kcontrols!)
- Now the volatile get should be working (I have no means to test it), put is working.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1706,6 +1709,11 @@ static int sof_ipc4_control_load_bytes(struct snd_sof_dev *sdev, struct snd_sof_
 	}
 
+	msg = &control_data->msg;
+	msg->primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
```
this is modifed in sof_ipc4_set_get_data() to SET/GET depending on whether it is a put/get. So maybe we dont need to se this here?

ujfalusi replied:
Yes, true. I copied the way how the volume control message is set up. It is easier to follow and understand by the reader and does not require knowledge that what will happen later on.

singalsu commented:
This version works perfectly with FIR and IIR in topology in https://github.com/thesofproject/sof/pull/6970 .

ranj063 commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -349,12 +349,6 @@ struct sof_ipc4_process {
 	void *ipc_config_data;
 	uint32_t ipc_config_size;
 	struct sof_ipc4_msg msg;
-	/*
```
@ujfalusi you missed one thing here. I think we should also remove the output_format field in this struct right?

ranj063 replied:
Actually @ujfalusi please don't remove the output_format. I think we will need it.

ranj063 reviewed (commented):
Looks great @ujfalusi. Just one comment abotu removing the out_format from struct sof_ipc4_process

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
+}
+
+static int _sof_ipc4_bytes_ext_get(struct snd_sof_control *scontrol,
+				  const unsigned int __user *binary_data,
```
there's an alignment issue here. Why doesnt checkpatch report it?

ujfalusi commented:
Changes since v5:
- parameter alignment fix in the last patch for _sof_ipc4_bytes_ext_get() (catched by @ranj063 and checkpatch --strict)

juimonen reviewed (dismissed):
looks good and logical series to me. Existing comments by previous reviewers that are not addressed, should be addressed/closed.

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
+	msg->extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(data->ipc4.param_id);
+
+	msg->data_ptr = data->data;
+	msg->data_size = data->size;
```
Hrm, interesting. We should be sending the sof_abi_hdr part as well to the firmware, not only the blob?

In IPC3 we have `struct sof_ipc_ctrl_data` which wraps all sorts of kcontrol related data (volume, enum, bytes), at the end of `struct sof_ipc_ctrl_data` we have a union and one option is the `sof_abi_hdr` (sof_ipc_ctrl_value_chan / sof_ipc_ctrl_value_comp / sof_abi_hdr)

With IPC4 we don't have such a generic kcontrol message, but we send everything with large_config and the target  module and it's parameter will tell the recipient what the data is and it will know how to handle it.

It is simple to send the header along, but it will need firmware side changes to handle the additional wrap (and btw, the abi_hdr use in firmware is just to `&cdata->data->data` to get to the real blob and the header's content ignored.

ujfalusi replied:
Right, after a quick check in my brain: it does not make sense to send the sof_abi_hdr along with the blobs. It could be only understood by SOF modules and IPC4 is much broader, we can not start sending blobs with strange headers to existing modules.

ujfalusi commented on include/uapi/sound/sof/header.h:
```diff
@@ -25,35 +31,21 @@
+			__u8 param_id;
+			__u8 reserved[3];
+		} ipc4;
+	};
```
we might not need this change at all?
Leaving the `type` as it were and document that it is param_id in IPC4.
Looking at the firmware code, the type is used for the same purpose as the param_id.

ujfalusi commented:
Changes since v6:
- Drop the modification of sof_abi_hdr, just document what the `type` is used in case of IPC4. At the end both IPC3 and IPC4 uses it for the same purpose, IPC3 sends it via IPC payload, IPC4 sends it via the header (param_id) and used as an ID of the blob to be used for.

juimonen reviewed (approved):
I'm fine with the abi_hdr change

kv2019i commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
+		return -EINVAL;
+	}
+
+	/* be->max is coming from topology */
```
Btw, this comment is also in IPC3 byte contorl implementation, but a bit hard to follow as  "be->max" is not used. "scontrol->max_size" is set based on "soc_bytes_ext", so I think this makes sense, but a bit hard to follow.

ujfalusi replied:
I will do a followup PR to update both IPC3 and IPC4, is that OK?

kv2019i reviewed (approved):
One minor "comment on comments", but this is not specific to this PR, so nothing blocking merge.

plbossart commented on include/uapi/sound/sof/header.h:
```diff
@@ -13,8 +13,12 @@
+ *	  for IPC3: Component specific type
+ *	  for IPC4: parameter ID (param_id) of the data
  * @size: The size in bytes of the data, excluding this struct
  * @abi: SOF ABI version
```
nit-pick: we should clarify that the ABI version is within the scope of the magic number, i.e. there is no relationship between the abi value for IPC3 and for IPC4.

ujfalusi replied:
Yes, it is a good idea, I will create a fixup for this.

ranj063 commented:
Mergin this so as to not lose approvals. @ujfalusi please submit fixups if needed
